### ISSUE TITLE
Chado storage again - The woes of too many joins

### ISSUE THREAD

**Issue by dsenalik** [tripal/tripal#1703]
### Tripal Version

4.x

### Drupal Version

10.1

### PostgreSQL Version

13

### PHP Version

8.2

### General Description

Chado storage is failing if we have too many joins, they may get added
to the SQL out of order. My example is I am working on a new field for
pub, and putting it on the study content type. This already has a
contact field, and both contact and pub have a join to the cvterm table.
Here is the sql generated by chado storage:
```
SELECT
"ct"."pub_id" AS "pub_id",
"ct"."contact_id" AS "contact_id",
"ct"."dbxref_id" AS "dbxref_id", 
"ct"."description" AS "description",
"ct"."name" AS "name",
"jpub0"."title" AS "jpub0_pub_title", 
"jpub0"."volumetitle" AS "jpub0_pub_volumetitle",
"jpub0"."volume" AS "jpub0_pub_volume", 
"jpub0"."series_name" AS "jpub0_pub_series_name",
"jpub0"."issue" AS "jpub0_pub_issue", 
"jpub0"."pyear" AS "jpub0_pub_pyear", 
"jpub0"."pages" AS "jpub0_pub_pages", 
"jpub0"."miniref" AS "jpub0_pub_miniref",
"jpub0"."uniquename" AS "jpub0_pub_uniquename", 
"jpub0"."is_obsolete" AS "jpub0_pub_is_obsolete", 
"jpub0"."publisher" AS "jpub0_pub_publisher", 
"jpub0"."pubplace" AS "jpub0_pub_pubplace",
"jcvterm0"."name" AS "jcvterm0_pub_type", 
"jcvterm1"."name" AS "jcvterm1_contact_type", 
"jcontact0"."name" AS "jcontact0_contact_name", 
"jcontact0"."description" AS "jcontact0_contact_description"
FROM chado.study "ct"
LEFT OUTER JOIN chado.pub "jpub0" ON ct.pub_id=jpub0.pub_id 
LEFT OUTER JOIN chado.cvterm "jcvterm0" ON jpub0.type_id=jcvterm0.cvterm_id 
   LEFT OUTER JOIN chado.cvterm "jcvterm1" ON jcontact0.type_id=jcvterm1.cvterm_id 
   LEFT OUTER JOIN chado.contact "jcontact0" ON ct.contact_id=jcontact0.contact_id
WHERE "ct"."study_id" = 1;
```
And note the two indented lines above. We have a failure because they are out of order,
we reference jcontact0 before this alias is defined. If I swap the order then the
query is fine.
```
   LEFT OUTER JOIN chado.contact "jcontact0" ON ct.contact_id=jcontact0.contact_id
   LEFT OUTER JOIN chado.cvterm "jcvterm1" ON jcontact0.type_id=jcvterm1.cvterm_id 
```

This happens because `addChadoRecordJoins()` adds each join into an array indexed
by both right table and parent column, rather than just adding them in order.
```
    // Add the join.
    $records[$parent_table][$delta]['joins'][$right_table][$parent_column]['on'] = [
      'left_table' => $left_table,
      'left_col' => $left_col,
      'right_table' => $right_table,
      'right_col' => $right_col,
      'left_alias' => $lalias,
      'right_alias' => $ralias,
```
The code to generate the SQL loops through the right table and sub-loops
through parent column. So all the cvterm joins get added together before the contact join.
![out-of-order-joins](https://github.com/tripal/tripal/assets/8419404/84a20a67-65e4-4ef1-950f-24145313d078)

We need to somehow preserve the join order...

### Steps to reproduce

1. Create a contact
1. Create a pub
1. Add the (from PR #1704) pub field to study content type
1. Create a study specifying both a contact and a pub.
![study-add-fail-screenshot](https://github.com/tripal/tripal/assets/8419404/4fb25504-5c04-4abe-b22a-4d920caf6299)

### Error messages

`PDOException: SQLSTATE[42P01]: Undefined table: 7 ERROR: missing FROM-clause entry for table "jcontact0" LINE 6: LEFT OUTER JOIN chado.cvterm "jcvterm1" ON jcontact0.type_id... ^ in /var/www/drupal/web/core/lib/Drupal/Core/Database/StatementWrapperIterator.php:110`
and so on...

**Comment by spficklin** [comment]
@dsenalik I would like to see if PR #1729 will fix this issue with your field.  I'm really hopeful it will fix the problem you were encountering.

**Comment by dsenalik** [comment]
Fixed with merged PR #1729
